### PR TITLE
Inception Package Improvements

### PR DESCRIPTION
**Pull request by qimingj**
- Implement inception cloud batch prediction. 
- Support explicit eval data in preprocessing.

### CODE REVIEW

File: datalab/mlalpha/_dataset.py [hunk 1]
```diff
@@ -46,6 +46,8 @@ def __init__(self, files, schema=None, schema_file=None):
     elif schema_file is not None:
       with ml.util._file.open_local_or_gcs(schema_file, 'r') as f:
         self._schema = json.load(f)
+    if isinstance(files, basestring):
```
Comment (brandondutra) [inline datalab/mlalpha/_dataset.py]: I just made a conflicting pull request for this class.

Basically, for the structured data package, I just want the file pattern.

File: solutionbox/inception/datalab_solutions/inception/_predictor.py [hunk 2]
```diff
@@ -82,57 +59,181 @@ def predict(model_dir, image_files):
...
+      targets = [x['label'] for x in element]
+
+    feed_dict = collections.defaultdict(list)
+    feed_dict[self._tf_inputs['image_bytes']] = [x['image_bytes'] for x in element]
+    feed_dict[self._tf_inputs['key']] = image_urls
+    predictions, labels, scores = self._session.run(
+        [self._tf_outputs['prediction'], self._tf_outputs['labels'], self._tf_outputs['scores']],
+        feed_dict=feed_dict)
+    if targets is not None:
+      yield zip(image_urls, targets, predictions, labels, scores)
```
Comment (brandondutra) [inline solutionbox/inception/datalab_solutions/inception/_predictor.py]: Do these zip's work as expected if feed_dict only has 1 element? Please test as I ran into that issue.

File: solutionbox/inception/datalab_solutions/inception/_predictor.py [hunk 3]
```diff
@@ -82,57 +59,181 @@ def predict(model_dir, image_files):
...
+    else:
+      yield zip(image_urls, predictions, labels, scores)
+
+
+class ProcessResultsDoFn(beam.DoFn):
+  """A DoFn that process prediction results."""
```
Comment (brandondutra) [inline solutionbox/inception/datalab_solutions/inception/_predictor.py]: better comment: process results how?

File: solutionbox/inception/datalab_solutions/inception/_predictor.py [hunk 4]
```diff
@@ -82,57 +59,181 @@ def predict(model_dir, image_files):
...
+    else:
+      csv.DictWriter(line, ['image_url', 'predicted', 'predicted_prob']).writerow(element)   
+    yield line.getvalue()
+
+
+def configure_pipeline(p, dataset, model_dir, output_csv, output_bq_table):
+  """Configures a dataflow pipeline for batch prediction."""
+
+  data = _util.get_sources_from_dataset(p, dataset, 'predict')
+  schema = _util.get_schema_from_dataset(dataset)
```
Comment (brandondutra) [inline solutionbox/inception/datalab_solutions/inception/_predictor.py]: Why not dataset.schema? Oh, you do some error checking. Still looks odd that you have to wrap "get schema" and our dataset classes cannot do this.

File: solutionbox/inception/datalab_solutions/inception/_util.py [hunk 5]
```diff
@@ -146,3 +146,56 @@ def accuracy(logits, labels):
...
+
+def get_schema_from_dataset(dataset):
+  """get schema from dataset."""
+
+  import csv
+  import datalab.bigquery as bq  
+  from datalab.mlalpha import CsvDataSet, BigQueryDataSet
+  import google.cloud.ml as ml
+
+  if type(dataset) is BigQueryDataSet:
```
Comment (brandondutra) [inline solutionbox/inception/datalab_solutions/inception/_util.py]: BigQueryDataSet.schema looks like it would be useful. Move this there?

File: solutionbox/inception/datalab_solutions/inception/_util.py [hunk 6]
```diff
@@ -146,3 +146,56 @@ def accuracy(logits, labels):
...
+  if type(dataset) is CsvDataSet:
+    source_list = []
+    for ii, input_path in enumerate(dataset.files):
+      source_list.append(p | 'Read from Csv %d (%s)' % (ii, mode) >> 
+          beam.io.ReadFromText(input_path, strip_trailing_newlines=True))
+    return (source_list | 'Flatten Sources (%s)' % mode >> beam.Flatten()
+        | 'Create Dict from Csv (%s)' % mode >>
+          beam.Map(lambda line: csv.DictReader([line], fieldnames=['image_url', 'label']).next()))
+  elif type(dataset) is BigQueryDataSet:
+    if len(dataset.sql.split()) == 1:
```
Comment (brandondutra) [inline solutionbox/inception/datalab_solutions/inception/_util.py]: We should refactor the BigQueryDataSet so it knows if it has a table or sql statement.

File: solutionbox/inception/datalab_solutions/inception/_util.py [hunk 7]
```diff
@@ -146,3 +146,56 @@ def accuracy(logits, labels):
...
+  import datalab.bigquery as bq  
+  from datalab.mlalpha import CsvDataSet, BigQueryDataSet
+  import google.cloud.ml as ml
+
+  if type(dataset) is BigQueryDataSet:
+    sql = dataset.sql
+    sql = 'SELECT * FROM (%s) LIMIT 1' % sql
+    return bq.Query(sql).results().schema
+  elif type(dataset) is CsvDataSet:
+    if dataset.schema is not None:
```
Comment (brandondutra) [inline solutionbox/inception/datalab_solutions/inception/_util.py]: how can this be none? This is the whole point to why we have csvdataset!  This brings up the point that class should enforce a schema value.

File: solutionbox/inception/datalab_solutions/inception/_predictor.py (hunk 2 above)
Comment (qimingj) [inline solutionbox/inception/datalab_solutions/inception/_predictor.py]: I tested with one instance and it works well. I did see your logic of special dealing with one instance. It it because TF returns scalar values instead of list if there is only one instance? For some reasons it works for me.

File: solutionbox/inception/datalab_solutions/inception/_predictor.py (hunk 3 above)
Comment (qimingj) [inline solutionbox/inception/datalab_solutions/inception/_predictor.py]: Done.

File: solutionbox/inception/datalab_solutions/inception/_util.py (hunk 7 above)
Comment (qimingj) [inline solutionbox/inception/datalab_solutions/inception/_util.py]: My initial thoughts about requiring schema or not in inception is that it seems boring for requiring users to give us: [{'name': 'image_url', 'type': 'STRING'}, {'name': 'label', 'type': 'STRING'}], where technically they should only be required to give CSV files with two columns. Users may struggle with namings such as 'image_url' vs 'imageurl', etc.

But I agree with you that forcing schema is not bad in long term, especially for BigQuery table we require tables with specified columns anyway. Will make schema required then.

File: solutionbox/inception/datalab_solutions/inception/_predictor.py (hunk 4 above)
Comment (qimingj) [inline solutionbox/inception/datalab_solutions/inception/_predictor.py]: Done.

File: solutionbox/inception/datalab_solutions/inception/_util.py (hunk 6 above)
Comment (qimingj) [inline solutionbox/inception/datalab_solutions/inception/_util.py]: Done. good idea.

File: solutionbox/inception/datalab_solutions/inception/_util.py (hunk 5 above)
Comment (qimingj) [inline solutionbox/inception/datalab_solutions/inception/_util.py]: Done.
